src/Stack/Package.hs

```diff
@@ -802,3 +810,46 @@ applyForceCustomBuild cabalVersion package
   forceCustomBuild =
        packageBuildType package == Simple
     && not (cabalVersion `withinRange` cabalVersionRange)
+
+-- | Check if the package has a main library that is buildable.
+hasMainBuildableLibrary :: Package -> Bool
+hasMainBuildableLibrary package =
+  maybe False isComponentBuildable $ packageLibrary package
+
+-- | Check if the main library has any exposed modules.
+--
+-- This should become irrelevant at some point since there's nothing inherently
+-- wrong or different with packages exposing only modules in internal libraries
+-- (for instance).
+mainLibraryHasExposedModules :: Package -> Bool
+mainLibraryHasExposedModules package =
+  maybe False (not . null . Component.exposedModules) $ packageLibrary package
+
+-- | Aggregate all unknown tools from all components. Mostly meant for
+-- build tools specified in the legacy manner (build-tools:) that failed the
+-- hard-coded lookup. See 'Stack.Types.Component.sbiUnknownTools' for more
+-- information.
+packageUnknownTools :: Package -> Set Text
+packageUnknownTools pkg = lib (bench <> tests <> flib <> sublib <> exe)
+ where
+  lib setT = case packageLibrary pkg of
+    Just libV -> addUnknownTools libV setT
+    Nothing -> setT
+  bench = gatherUnknownTools $ packageBenchmarkSuites pkg
+  tests = gatherUnknownTools $ packageTestSuites pkg
+  flib = gatherUnknownTools $ packageForeignLibraries pkg
+  sublib = gatherUnknownTools $ packageSubLibraries pkg
+  exe = gatherUnknownTools $ packageExecutables pkg
+  addUnknownTools :: HasBuildInfo x => x -> Set Text -> Set Text
+  addUnknownTools = (<>) . Component.sbiUnknownTools . getField @"buildInfo"
+  gatherUnknownTools :: HasBuildInfo x => CompCollection x -> Set Text
+  gatherUnknownTools = foldr' addUnknownTools mempty
+
+packageSubLibrariesNameSet :: Package -> Set Text
```

Comment:
I wondered if this would be better renamed `buildableSubLibs` (see my similar comment below for explanation).

Also, good to have `buildableForeignLibs` and `buildableTestSuites`, to complete the set of similar functions.

Reply:
I have given effect to the first part of this suggestion (`buildableSubLibs`).

Reply:
I have also given effect to new `Stack.Package.buildableTestSuites`.

Reply:
Finally, I have also given effect to new `Stack.Package.buildableForeignLibs`.